api/workflow/v1alpha1/types.go

```diff
@@ -306,6 +325,11 @@ func (n NodeStatus) Successful() bool {
 	return n.Phase == NodeSucceeded || n.Phase == NodeSkipped
 }
 
+// Failed returns whether the node failed during execution.
+func (n NodeStatus) Failed() bool {
+	return n.Phase == NodeFailed
```

Comment:
Would node deletion be treated as Failure?

Reply:
Node deletion currently falls into the category of Error. My opinion is that yes it should be retried.

Reply:
However this prompted me to think that we may need to distinguish between unretryable errors (e.g. the artifact does not exist, the secret does not exist). Perhaps this is why this code is only considering Failed to retry.

Reply:
Depends on what problem this is trying to solve. The current implementation is mainly focussed on retrying when some program logic or dependencies internal to the pod don't work out. As you point out, if we want to retry on factors other than this, there is a need for a more sophisticated logic.

Instead of simply looking at Failed(), let me create a method called `ShouldRetry()`. This can encapsulate the logic for distinguishing between various error types and according simply returning whether or not the pod should be run again.